### ISSUE TITLE
pack_padded_sequence output is not as expected

### ISSUE THREAD

**Issue by stefbraun** [pytorch/pytorch#1820]
Hi, ```pack_padded_sequence``` creates a ```Packed Sequence``` object with ```(data, batch_sizes)```. The way how ```data``` is created does not meet (standard?) expectations: instead of concatenating samples from the batch without padding, it seems to do something else, see the plot below.

In fact, if I were to run a fully connected layer on ```data``` - e.g. to map a RNN output to classes when using CTC - the structure of ```data``` would not make sense. I have a script that produces a plot to showcase the behaviour. Can you comment and possibly clarify the documentation if this is not a bug?
![issue](https://user-images.githubusercontent.com/13469638/27219641-b7b8cbec-5282-11e7-8a57-805d149792f2.png)

```
import torch
import numpy as np
import matplotlib.pyplot as plt
from torch.nn.utils.rnn import pack_padded_sequence, pad_packed_sequence
from torch.autograd import Variable

# Get data
single_sample = np.array([[1, 2, 3, 0, 0], [1, 2, 3, 0, 0], [1, 2, 3, 0, 0], [1, 2, 3, 0, 0]], dtype='float32')  # (features, max_len)
single_sample = single_sample.T  # (max_len, features)
batch = np.array([single_sample] * 3)  # (batch_size, max_len, features)
lengths = np.array([3, 3, 3])

# Move to pytorch
batch_pt = Variable(torch.from_numpy(batch))

# Pack the sequence
xp = pack_padded_sequence(batch_pt, lengths, batch_first=True)

# Expected output of xp.data
xp_exp = np.array([single_sample[:3, :]] * 3)  # (batch_size, max_len, features)
xp_exp = np.reshape(xp_exp, (9, -1))

# Plot
fig, (ax1, ax2, ax3) = plt.subplots(3)
# ax1
im1 = ax1.imshow(single_sample.T, aspect='auto', interpolation='none', vmin=0, vmax=3)
fig.colorbar(im1, ax=ax1)
ax1.set_title('Batch contains three of these single samples')
ax1.set_xlabel('Time')
ax1.set_ylabel('Features')
# ax2
im2 = ax2.imshow(xp.data.data.numpy().T, aspect='auto', interpolation='none', vmin=0, vmax=3)
fig.colorbar(im2, ax=ax2)
ax2.set_title('Pytorch packed_sequence.data')
ax2.set_xlabel('Time')
ax2.set_ylabel('Features')
# ax3
im3 = ax3.imshow(xp_exp.T, aspect='auto', interpolation='none', vmin=0, vmax=3)
fig.colorbar(im3, ax=ax3)
ax3.set_title('Expected packed_sequence.data')
ax3.set_xlabel('Time')
ax3.set_ylabel('Features')
# Other
plt.tight_layout()
plt.show()
```

**Comment by jekbradbury** [comment]
This is expected. The (unusual) format of a `PackedSequence` object is the way it is because of the API of cuDNN's RNN/LSTM functions, which expect a time-sequence of variable-length batches rather than a batch-sequence of variable-length samples. If your samples are `a`, `bc`, and `defg` the `PackedSequence` will contain the packed tensor `abdcefg` and a list of lengths `[3, 2, 1, 1]`.

**Comment by stefbraun** [comment]
Thanks a lot for the clarification! After rethinking, a dense layer on ```data``` should still work, as it only sees single frames at a time, so the ordering of frames is not important.
